dask/dataframe/core.py

```diff
@@ -309,6 +310,261 @@ def _scalar_binary(op, self, other, inv=False):
+    "Lazy" statistics correspond to partition statistics
+    that can be loaded "just in time" using a callback
+    function. The call-back function must accept an
+    optional `keys: set` argument to specify which columns
+    to load statistics for.
+    """
+
+    _statistics: dict
+
+    def __init__(
+        self,
+        statistics: dict | None = None,
+    ):
+        self._statistics = statistics or {}
+
+    def copy(self, keys: set | None = None) -> PartitionStatistics:
+        """Return a copy of this PartitionStatistics object"""
+        if keys is not None:
+            return PartitionStatistics({k: self._statistics[k] for k in keys})
+        return PartitionStatistics(self._statistics.copy())
+
+    @property
+    def known_stats(self) -> set:
+        """Return known (in memory) partition-statistic keys"""
+        return {k for k, v in self._statistics.items() if not callable(v)}
+
+    @property
+    def lazy_stats(self) -> set:
+        """Return lazy partition-statistic keys"""
+        return {k for k, v in self._statistics.items() if callable(v)}
+
+    @property
+    def available_stats(self) -> set:
+        """Return all available partition-statistic keys"""
+        return self.known_stats | self.lazy_stats
+
+    def get(self, keys: set) -> dict:
+        """Return requested partition statistics"""
+        success = self.load_lazy_stats(keys)
+        if not success:
+            missing = self.known_stats - keys
+            raise KeyError(f"{missing} stats could not be loaded")
+        return {k: self._statistics[k] for k in keys}
+
+    def load_lazy_stats(self, keys: set | None = None) -> bool:
+        """Load lazy statistics
+
+        Returns value specifies if the desired ``keys``
+        were successfully added to the known statistics.
+        """
+        _available = self.available_stats
+        if keys:
+            if not keys.issubset(_available):
+                raise KeyError(
+                    f"{keys} stat keys were requested, "
+                    f"but only {_available} stats are available."
+                )
+        else:
+            keys = _available
+
+        _keys = keys - self.known_stats
+        if not _keys:
+            return True  # No lazy stats to load
+
+        # Aggregate lazy-function calls
+        _lazy_stats = defaultdict(set)
+        for k in _keys:
+            _func = self._statistics[k]
+            _lazy_stats[_func].add(k)
+
+        # Perform update
+        new_stats = {}
+        for _func, _keyset in _lazy_stats.items():
+            new_stats.update(_func(keys=_keyset))
+        self._statistics.update(new_stats)
+
+        # Check if update was successful
+        if _keys - set(new_stats.keys()):
+            return False
+        return True
+
+
+class PartitionMetadata:
+    """
+    Container for DataFrame partition metadata
+    """
+
+    _meta: Any
+    _npartitions: int
+    _divisions: tuple | None
+    _partitioning: dict
```

Comment:
I think part of what I'm reacting to is that there are seemingly many ways for `npartitions`, `divisions`, `partitioning`, and `statistics` to be mutually inconsistent. There are a number of checks that you are doing in the constructor and elsewhere to avoid that, but they are likely not exhaustive.

I want to find a way to represent all of this data in such way that there is only one source of truth for "information about metadata and partitioning", and all of the older representations like `npartitions`, `divisions`, etc would just be thin views or properties on top of that.